File: LeetCode/210.Course_Schedule_II/210.Course_Schedule_II.py
#!/usr/bin/env python3


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # bi -> ai
        # topological sort kahn algo

        # create graph based on prerequisite
        graph = defaultdict(list)
        indegree = [0] * numCourses

        for p in prerequisites:
            a, b = p[0], p[1]
            graph[b].append(a)
            indegree[a] += 1

        # bfs
        q = deque([i for i in range(numCourses) if indegree[i] == 0])
        res = []

        while q:
            e = q.popleft()
            res.append(e)

            for nei in graph[e]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)

        return res if len(res) == numCourses else []

File: LeetCode/210.Course_Schedule_II/210.Course_Schedule_II.swift
class Solution {
    func findOrder(_ numCourses: Int, _ prerequisites: [[Int]]) -> [Int] {
        // [0, numCourses-1]
        // [ai, bi] => bi -> ai
        // Dijstra

        var indeg: [Int] = Array(repeating: 0, count: numCourses)
        var graph: [Int: [Int]] = [:]

        for pre in prerequisites {
            let (a, b) = (pre[0], pre[1])
            graph[b, default: []].append(a)
            indeg[a] += 1
        }

        var res: [Int] = []
        var head = 0
        for (i, v) in indeg.enumerated() {
            if v == 0 {
                res.append(i)
            }
        }

        while head < res.count {
            let v = res[head]
            head += 1
            for nei in graph[v, default: []] {
                indeg[nei] -= 1
                if indeg[nei] == 0 {
                    res.append(nei)
                }
            }
        }

        return numCourses > res.count ? [] : res
    }
}
